File: python3/maximum_area_of_a_piece_of_cake_after_horizontal_and_vertical_cuts.py
"""
Space : O(n)
Time  : O(n log n)
"""


class Solution:
    def maxArea(self, h: int, w: int, horizontalCuts: List[int], verticalCuts: List[int]) -> int:
        hc = [0] + sorted(horizontalCuts) + [h]
        vc = [0] + sorted(verticalCuts) + [w]
        mh = mv = 0

        for i in range(1, len(hc)):
            mh = max(mh, hc[i] - hc[i-1])

        for i in range(1, len(vc)):
            mv = max(mv, vc[i] - vc[i-1])

        return (mh * mv) % (10**9 + 7)

File: golang/maximum_area_of_a_piece_of_cake_after_horizontal_and_vertical_cuts.go
package main

// Space : O(1)
// Time  : O(n log n)

func maxArea(h int, w int, horizontalCuts []int, verticalCuts []int) int {
    horizontalCuts = append(horizontalCuts, 0)
    horizontalCuts = append(horizontalCuts, h)
    verticalCuts = append(verticalCuts, 0)
    verticalCuts = append(verticalCuts, w)
    
    sort.Ints(horizontalCuts)
    sort.Ints(verticalCuts)
    
    mh := 0
    for i := 1; i < len(horizontalCuts); i++ {
        mh = Max(mh, horizontalCuts[i] - horizontalCuts[i-1])
    }
    
    mv := 0
    for j := 1; j < len(verticalCuts); j++ {
        mv = Max(mv, verticalCuts[j] - verticalCuts[j-1])
    }
    
    return (mh * mv) % (int(math.Pow(10, 9)) + 7)